presto-spi/src/main/java/com/facebook/presto/spi/function/ScalarFunctionImplementation.java

```diff
@@ -0,0 +1,29 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.function;
+
+import com.facebook.presto.spi.api.Experimental;
+
+@Experimental
+public interface ScalarFunctionImplementation
+{
+    /**
+     * We use FunctionImplementationType from FunctionMetadata to know the actual type of implementation object.
+     * The FunctionImplementationType -> Object type mapping is:
+     * BUILTIN -> BuiltInScalarFunctionImplementation
+     * SQL     -> String
+     */
+
+    Object getImplementation();
```

Comment:
This design pattern doesn't look at an "Interface design pattern" but more like a "delegation". `FunctionImplementationType` should be good enough for caller to decide who to delegate. This interface may not be necessary. In the worst case, presto-main depends on presto-sql-function, `parseSqlFunctionExpression` can down cast it:
```java
((SqlInvokedRegularSqlFunctionImplementation) functionImplementation).getImplementation()
```
where only `SqlInvokedRegularSqlFunctionImplementation` needs a `String getImplementation()` method.